File: scala3/src/main/scala/io/github/sps23/designpatterns/builder/HttpClientConfig.scala
package io.github.sps23.designpatterns.builder

final case class HttpClientConfig(
    host: String,
    port: Int,
    connectTimeoutMs: Int     = 500,
    readTimeoutMs: Int        = 2000,
    useSsl: Boolean           = true,
    maxRetries: Int           = 3,
    retryBackoffMs: List[Int] = List(100, 200, 500),
    defaultHeaders: Map[String, String] = Map(
      "Accept"     -> "application/json",
      "User-Agent" -> "java-for-scala-devs-client"
    ),
    circuitBreakerFailureThreshold: Int = 50,
    apiVersion: String                  = "v1",
    enableCompression: Boolean          = true
):
  def baseUrl: String =
    val scheme = if useSsl then "https" else "http"
    s"$scheme://$host:$port/api/$apiVersion"

object HttpClientConfigBuilder:
  def builder(host: String, port: Int): HttpClientConfigBuilder =
    HttpClientConfigBuilder(host = host, port = port)

final case class HttpClientConfigBuilder private (
    host: String,
    port: Int,
    connectTimeoutMs: Int     = 500,
    readTimeoutMs: Int        = 2000,
    useSsl: Boolean           = true,
    maxRetries: Int           = 3,
    retryBackoffMs: List[Int] = List(100, 200, 500),
    defaultHeaders: Map[String, String] = Map(
      "Accept"     -> "application/json",
      "User-Agent" -> "java-for-scala-devs-client"
    ),
    circuitBreakerFailureThreshold: Int = 50,
    apiVersion: String                  = "v1",
    enableCompression: Boolean          = true
):
  def withConnectTimeoutMs(value: Int): HttpClientConfigBuilder =
    copy(connectTimeoutMs = value)

  def withReadTimeoutMs(value: Int): HttpClientConfigBuilder =
    copy(readTimeoutMs = value)

  def withUseSsl(value: Boolean): HttpClientConfigBuilder =
    copy(useSsl = value)

  def withMaxRetries(value: Int): HttpClientConfigBuilder =
    copy(maxRetries = value)

  def withRetryBackoffMs(values: List[Int]): HttpClientConfigBuilder =
    copy(retryBackoffMs = values)

  def withDefaultHeaders(values: Map[String, String]): HttpClientConfigBuilder =
    copy(defaultHeaders = values)

  def addDefaultHeader(key: String, value: String): HttpClientConfigBuilder =
    copy(defaultHeaders = defaultHeaders + (key -> value))

  def withCircuitBreakerFailureThreshold(value: Int): HttpClientConfigBuilder =
    copy(circuitBreakerFailureThreshold = value)

  def withApiVersion(value: String): HttpClientConfigBuilder =
    copy(apiVersion = value)

  def withEnableCompression(value: Boolean): HttpClientConfigBuilder =
    copy(enableCompression = value)

  def build(): HttpClientConfig =
    validateHost(host)
    validatePort(port)
    validateTimeout(connectTimeoutMs, "Connect timeout")
    validateTimeout(readTimeoutMs, "Read timeout")
    validateTimeoutRelationship(connectTimeoutMs, readTimeoutMs)
    validateMaxRetries(maxRetries)
    validateBackoff(retryBackoffMs, maxRetries)
    validateHeaders(defaultHeaders)
    validateThreshold(circuitBreakerFailureThreshold)
    validateApiVersion(apiVersion)
    HttpClientConfig(
      host,
      port,
      connectTimeoutMs,
      readTimeoutMs,
      useSsl,
      maxRetries,
      retryBackoffMs,
      defaultHeaders,
      circuitBreakerFailureThreshold,
      apiVersion,
      enableCompression
    )

  private def validateHost(value: String): Unit =
    if value == null || value.trim.isEmpty then
      throw new IllegalArgumentException("Host must not be blank")

  private def validatePort(value: Int): Unit =
    if value < 1 || value > 65535 then
      throw new IllegalArgumentException("Port must be between 1 and 65535")

  private def validateTimeout(value: Int, name: String): Unit =
    if value <= 0 then throw new IllegalArgumentException(s"$name must be positive")

  private def validateTimeoutRelationship(connectValue: Int, readValue: Int): Unit =
    if readValue < connectValue then
      throw new IllegalArgumentException(
        "Read timeout must be greater than or equal to connect timeout"
      )

  private def validateMaxRetries(value: Int): Unit =
    if value < 0 || value > 10 then
      throw new IllegalArgumentException("Max retries must be between 0 and 10")

  private def validateBackoff(values: List[Int], retries: Int): Unit =
    if values == null then throw new IllegalArgumentException("Retry backoff must not be null")
    else if values.size != retries then
      throw new IllegalArgumentException("Retry backoff size must match max retries")
    else if values.exists(_ <= 0) then
      throw new IllegalArgumentException("Retry backoff values must be positive")

  private def validateHeaders(values: Map[String, String]): Unit =
    if values == null then throw new IllegalArgumentException("Default headers must not be null")
    else
      values.foreach {
        case (key, _) if key == null || key.trim.isEmpty =>
          throw new IllegalArgumentException("Header key must not be blank")
        case (_, headerValue) if headerValue == null || headerValue.trim.isEmpty =>
          throw new IllegalArgumentException("Header value must not be blank")
        case _ =>
      }

  private def validateThreshold(value: Int): Unit =
    if value < 1 || value > 100 then
      throw new IllegalArgumentException("Circuit breaker threshold must be between 1 and 100")

  private def validateApiVersion(value: String): Unit =
    if value == null || !value.matches("v\\d+") then
      throw new IllegalArgumentException("API version must match pattern v{number}")

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/builder/HttpClientConfig.kt
package io.github.sps23.designpatterns.builder

data class HttpClientConfig(
    val host: String,
    val port: Int,
    val connectTimeoutMs: Int = 500,
    val readTimeoutMs: Int = 2000,
    val useSsl: Boolean = true,
    val maxRetries: Int = 3,
    val retryBackoffMs: List<Int> = listOf(100, 200, 500),
    val defaultHeaders: Map<String, String> =
        mapOf(
            "Accept" to "application/json",
            "User-Agent" to "java-for-scala-devs-client",
        ),
    val circuitBreakerFailureThreshold: Int = 50,
    val apiVersion: String = "v1",
    val enableCompression: Boolean = true,
) {
    val baseUrl: String
        get() {
            val scheme = if (useSsl) "https" else "http"
            return "$scheme://$host:$port/api/$apiVersion"
        }
}

class HttpClientConfigBuilder private constructor(
    private val host: String,
    private val port: Int,
) {
    private var connectTimeoutMs: Int = 500
    private var readTimeoutMs: Int = 2000
    private var useSsl: Boolean = true
    private var maxRetries: Int = 3
    private var retryBackoffMs: List<Int> = listOf(100, 200, 500)
    private val defaultHeaders: MutableMap<String, String> =
        linkedMapOf(
            "Accept" to "application/json",
            "User-Agent" to "java-for-scala-devs-client",
        )
    private var circuitBreakerFailureThreshold: Int = 50
    private var apiVersion: String = "v1"
    private var enableCompression: Boolean = true

    fun connectTimeoutMs(value: Int): HttpClientConfigBuilder {
        connectTimeoutMs = value
        return this
    }

    fun readTimeoutMs(value: Int): HttpClientConfigBuilder {
        readTimeoutMs = value
        return this
    }

    fun useSsl(value: Boolean): HttpClientConfigBuilder {
        useSsl = value
        return this
    }

    fun maxRetries(value: Int): HttpClientConfigBuilder {
        maxRetries = value
        return this
    }

    fun retryBackoffMs(values: List<Int>): HttpClientConfigBuilder {
        retryBackoffMs = values
        return this
    }

    fun defaultHeaders(values: Map<String, String>): HttpClientConfigBuilder {
        defaultHeaders.clear()
        defaultHeaders.putAll(values)
        return this
    }

    fun addDefaultHeader(
        key: String,
        value: String,
    ): HttpClientConfigBuilder {
        defaultHeaders[key] = value
        return this
    }

    fun circuitBreakerFailureThreshold(value: Int): HttpClientConfigBuilder {
        circuitBreakerFailureThreshold = value
        return this
    }

    fun apiVersion(value: String): HttpClientConfigBuilder {
        apiVersion = value
        return this
    }

    fun enableCompression(value: Boolean): HttpClientConfigBuilder {
        enableCompression = value
        return this
    }

    fun build(): HttpClientConfig {
        require(host.isNotBlank()) { "Host must not be blank" }
        require(port in 1..65535) { "Port must be between 1 and 65535" }
        require(connectTimeoutMs > 0) { "Connect timeout must be positive" }
        require(readTimeoutMs > 0) { "Read timeout must be positive" }
        require(readTimeoutMs >= connectTimeoutMs) {
            "Read timeout must be greater than or equal to connect timeout"
        }
        require(maxRetries in 0..10) { "Max retries must be between 0 and 10" }
        require(retryBackoffMs.size == maxRetries) { "Retry backoff size must match max retries" }
        require(retryBackoffMs.all { it > 0 }) { "Retry backoff values must be positive" }
        require(circuitBreakerFailureThreshold in 1..100) {
            "Circuit breaker threshold must be between 1 and 100"
        }
        require(Regex("v\\d+").matches(apiVersion)) {
            "API version must match pattern v{number}"
        }
        require(defaultHeaders.all { it.key.isNotBlank() }) { "Header key must not be blank" }
        require(defaultHeaders.all { it.value.isNotBlank() }) { "Header value must not be blank" }

        return HttpClientConfig(
            host = host,
            port = port,
            connectTimeoutMs = connectTimeoutMs,
            readTimeoutMs = readTimeoutMs,
            useSsl = useSsl,
            maxRetries = maxRetries,
            retryBackoffMs = retryBackoffMs.toList(),
            defaultHeaders = defaultHeaders.toMap(),
            circuitBreakerFailureThreshold = circuitBreakerFailureThreshold,
            apiVersion = apiVersion,
            enableCompression = enableCompression,
        )
    }

    companion object {
        fun builder(
            host: String,
            port: Int,
        ): HttpClientConfigBuilder = HttpClientConfigBuilder(host, port)
    }
}
